File: mplad-app/lib/pipeline/detectors/d6.js
async function runD6(works, db) {
  console.log('[D6] Running Geo (Ghost & Overlap) Detector...');
  
  const results = [];

  // Haversine distance in km
  function haversine(lat1, lon1, lat2, lon2) {
    if (!lat1 || !lon1 || !lat2 || !lon2) return Infinity;
    const R = 6371; // Earth radius in km
    const dLat = (lat2 - lat1) * Math.PI / 180;
    const dLon = (lon2 - lon1) * Math.PI / 180;
    const a = 
      Math.sin(dLat/2) * Math.sin(dLat/2) +
      Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) * 
      Math.sin(dLon/2) * Math.sin(dLon/2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
    return R * c;
  }

  // 1. Ghost Distance
  // Fetch all villages and group by district
  const allVillages = await db.village.findMany();
  const villagesByDistrict = {};
  for (const v of allVillages) {
    if (!villagesByDistrict[v.district_id]) villagesByDistrict[v.district_id] = [];
    villagesByDistrict[v.district_id].push(v);
  }

  // Fetch fund flows for spend ratio
  const fundFlows = await db.fundFlow.findMany();
  const spentByDistrict = {};
  const entitlementByDistrict = {};
  for (const f of fundFlows) {
    if (!spentByDistrict[f.district_id]) spentByDistrict[f.district_id] = 0;
    if (!entitlementByDistrict[f.district_id]) entitlementByDistrict[f.district_id] = 0;
    spentByDistrict[f.district_id] += (f.expenditure || 0);
    entitlementByDistrict[f.district_id] += (f.entitlement || 0);
  }

  for (const w of works) {
    if (!w.district_id || !w.lat || !w.lon) continue;
    
    const vils = villagesByDistrict[w.district_id] || [];
    let minDist = Infinity;
    let nearestVil = null;

    for (const v of vils) {
      if (v.lat && v.lon) {
        const d = haversine(w.lat, w.lon, v.lat, v.lon);
        if (d < minDist) {
          minDist = d;
          nearestVil = v.id;
        }
      }
    }

    if (minDist !== Infinity && minDist >= 3.0) {
      let spend_ratio = 0;
      if (entitlementByDistrict[w.district_id] > 0) {
        spend_ratio = spentByDistrict[w.district_id] / entitlementByDistrict[w.district_id];
      }
      
      results.push({
        work_id: w.id,
        detector: 'D6',
        score: minDist, // raw distance
        evidence_json: JSON.stringify({
          distance_km: minDist,
          nearest_village: nearestVil,
          status: w.status,
          spend_ratio: spend_ratio
        })
      });
    }
  }

  // 2. DBSCAN clustering per district for Overlap
  const eps_km = 0.3; // 300m
  const min_samples = 2;

  const worksByDistrict = {};
  for (const w of works) {
    if (w.district_id && w.lat && w.lon) {
      if (!worksByDistrict[w.district_id]) worksByDistrict[w.district_id] = [];
      worksByDistrict[w.district_id].push(w);
    }
  }

  for (const [distId, distWorks] of Object.entries(worksByDistrict)) {
    const visited = new Set();
    const clusters = [];
    
    function getNeighbors(idx) {
      const neighbors = [];
      const w1 = distWorks[idx];
      for (let i = 0; i < distWorks.length; i++) {
        const w2 = distWorks[i];
        if (haversine(w1.lat, w1.lon, w2.lat, w2.lon) <= eps_km) {
          neighbors.push(i);
        }
      }
      return neighbors;
    }

    for (let i = 0; i < distWorks.length; i++) {
      if (visited.has(i)) continue;
      visited.add(i);

      const neighbors = getNeighbors(i);
      if (neighbors.length < min_samples) {
        // Noise
        continue;
      }

      const cluster = [i];
      let seedIdx = 0;
      while (seedIdx < neighbors.length) {
        const p = neighbors[seedIdx];
        if (!visited.has(p)) {
          visited.add(p);
          const pNeighbors = getNeighbors(p);
          if (pNeighbors.length >= min_samples) {
            neighbors.push(...pNeighbors);
          }
        }
        if (!cluster.includes(p)) {
          cluster.push(p);
        }
        seedIdx++;
      }
      clusters.push(cluster.map(idx => distWorks[idx]));
    }

    // Process clusters for overlap
    for (const cluster of clusters) {
      // Find works with same category and same FY
      const catFyGroups = {};
      for (const w of cluster) {
        const key = `${w.category}|${w.fy}`;
        if (!catFyGroups[key]) catFyGroups[key] = [];
        catFyGroups[key].push(w);
      }

      for (const [key, groupWorks] of Object.entries(catFyGroups)) {
        if (groupWorks.length > 1) { // Same cluster + same category + same FY
          const clusterMembers = groupWorks.map(gw => gw.id);
          const cats = groupWorks.map(gw => gw.category);
          
          for (const gw of groupWorks) {
            results.push({
              work_id: gw.id,
              detector: 'D6_overlap',
              score: 1.0,
              evidence_json: JSON.stringify({
                cluster_members: clusterMembers,
                categories: cats
              })
            });
          }
        }
      }
    }
  }

  return { results, districtFlags: [] };
}

module.exports = { runD6 };

File: mplad-app/app/pipeline/detectors/d6.py
import sqlite3
import pandas as pd
import numpy as np
import json
import yaml
from sklearn.cluster import DBSCAN
from app.pipeline.detectors.d1 import haversine

def run_d6(db_path):
    print("Running D6 (Geo)...")
    con = sqlite3.connect(db_path)
    
    with open("config/detectors.yaml", "r") as f:
        conf = yaml.safe_load(f)
        
    storage_floor = conf.get("d6", {}).get("storage_floor_km", 3.0)
    
    works_df = pd.read_sql_query("SELECT id, district_id, lat, lon, status, sanctioned_amount, expenditure, category, fy, village FROM works", con)
    villages_df = pd.read_sql_query("SELECT id, district_id, name, lat as centroid_lat, lon as centroid_lon FROM villages", con)
    
    if works_df.empty:
        con.close()
        return
        
    detection_rows = []
    
    for dist_id, dist_works in works_df.groupby("district_id"):
        dist_villages = villages_df[villages_df.district_id == dist_id]
        
        # DBSCAN overlap logic
        # Convert lat/lon to radians for haversine
        coords = np.radians(dist_works[['lat', 'lon']].values)
        if len(coords) >= 2:
            eps_rad = 300 / (6371 * 1000) # 300 meters
            db = DBSCAN(eps=eps_rad, min_samples=2, metric='haversine').fit(coords)
            dist_works = dist_works.copy()
            dist_works['cluster'] = db.labels_
            
            # Find clusters with same category and FY overlap +/- 1
            for cluster_id, cluster_group in dist_works[dist_works['cluster'] != -1].groupby('cluster'):
                if len(cluster_group) >= 2:
                    # Check pairs within the cluster for overlap
                    # O(n^2) but n is small (points within 300m)
                    rows = cluster_group.to_dict("records")
                    caught_ids = set()
                    
                    for i in range(len(rows)):
                        for j in range(i+1, len(rows)):
                            w1 = rows[i]
                            w2 = rows[j]
                            
                            if w1['category'] == w2['category']:
                                fy1 = int(w1['fy'][:4])
                                fy2 = int(w2['fy'][:4])
                                if abs(fy1 - fy2) <= 1:
                                    caught_ids.add(w1['id'])
                                    caught_ids.add(w2['id'])
                                    
                    # Emit D6_overlap for all caught works in this cluster
                    if caught_ids:
                        member_ids = sorted(list(caught_ids))
                        cats = sorted(list(set(r['category'] for r in rows if r['id'] in caught_ids)))
                        
                        evidence = {
                            "cluster_members": member_ids,
                            "categories": cats
                        }
                        
                        for wid in caught_ids:
                            detection_rows.append((
                                wid,
                                "D6_overlap",
                                1.0,
                                json.dumps(evidence)
                            ))
                            
        # Ghost distance logic
        if not dist_villages.empty:
            v_lats = dist_villages['centroid_lat'].values
            v_lons = dist_villages['centroid_lon'].values
            v_names = dist_villages['name'].values
            
            for _, w in dist_works.iterrows():
                # vectorize haversine for this work against all villages
                # formula from d1.py haversine:
                # it's just a loop, let's just do a loop since it's simple enough
                own_village = dist_villages[dist_villages['name'] == w['village']]
                if not own_village.empty:
                    v = own_village.iloc[0]
                    nearest_village = v['name']
                    min_dist = haversine(w['lat'], w['lon'], v['centroid_lat'], v['centroid_lon'])
                else:
                    min_dist = float('inf')
                    nearest_village = None
                    for v_lat, v_lon, v_name in zip(v_lats, v_lons, v_names):
                        d = haversine(w['lat'], w['lon'], v_lat, v_lon)
                        if d < min_dist:
                            min_dist = d
                            nearest_village = v_name
                        
                if min_dist >= storage_floor:
                    # spend ratio
                    exp = w['expenditure'] if pd.notna(w['expenditure']) else 0.0
                    sanc = w['sanctioned_amount'] if pd.notna(w['sanctioned_amount']) and w['sanctioned_amount'] > 0 else 1.0
                    spend_ratio = exp / sanc
                    
                    evidence = {
                        "distance_km": float(min_dist),
                        "nearest_village": nearest_village,
                        "status": w['status'],
                        "spend_ratio": float(spend_ratio),
                        "own_village_distance": float(min_dist)
                    }
                    detection_rows.append((
                        w['id'],
                        "D6",
                        float(min_dist),
                        json.dumps(evidence)
                    ))
                    
    con.executemany("INSERT INTO detection_results (work_id, detector, score, evidence_json) VALUES (?, ?, ?, ?)", detection_rows)
    con.commit()
    con.close()
